lib/rubygems/printable_uri.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require 'uri'
```

Comment:
Do we need to vendor https://github.com/ruby/uri now as well?

Reply:
I'm not fully sure, but `Gem::UriParser` requires `uri` lazily and things seem fine, so we should probably keep that pattern and move the require to be lazy here too.

We could also avoid using `uri` entirelly in this class by changing `valid_uri?` implementation to `!@uri.is_a?(String)`? If we do that, maybe `parsed_as_valid_uri?` would be a better name? Not sure.

Reply:
I have removed it. please take a look at this commit: [`04dc862` (#4868)](https://github.com/rubygems/rubygems/pull/4868/commits/04dc862f80c99c4ca599156c046cea18dd98fad8)

Reply:
Great!

Reply:
I'm trying to figure out if this could be better:

```diff
@@ -20,10 +20,7 @@ def parse_uri
   end
 
   def valid_uri?
-    @uri.respond_to?(:user) &&
-      @uri.respond_to?(:user=) &&
-      @uri.respond_to?(:password) &&
-      @uri.respond_to?(:password=)
+    [:user, :user=, :password, :password=].all? { |method| @uri.respond_to?(method) }
   end

```

Thoughts?

Reply:
\cc @deivid-rodriguez